src/c-unchop.c

```diff
@@ -82,88 +82,74 @@ static SEXP vec_unchop(SEXP x,
     return R_NilValue;
   }
 
-  x = PROTECT(vec_cast_common(x, ptype));
+  xs = PROTECT(vec_cast_common(xs, ptype));
 
-  SEXP x_names = PROTECT(r_names(x));
-
-  bool has_outer_names = (x_names != R_NilValue);
   bool assign_names = !Rf_inherits(name_spec, "rlang_zap");
-  bool has_names =
-    assign_names &&
-    !is_data_frame(ptype) &&
-    (has_outer_names || list_has_inner_vec_names(x, x_size));
-
-  // Element sizes are only required for applying the `name_spec`
-  SEXP sizes = vctrs_shared_empty_int;
-  if (has_names) {
-    sizes = Rf_allocVector(INTSXP, x_size);
-  }
-  PROTECT(sizes);
-  int* p_sizes = INTEGER(sizes);
+  SEXP xs_names = PROTECT(r_names(xs));
+  bool xs_is_named = xs_names != R_NilValue && !is_data_frame(ptype);
 
   R_len_t out_size = 0;
 
   // `out_size` is computed from `indices`
   for (R_len_t i = 0; i < x_size; ++i) {
-    SEXP elt = VECTOR_ELT(x, i);
+    SEXP x = VECTOR_ELT(xs, i);
 
-    if (elt == R_NilValue) {
+    if (x == R_NilValue) {
       continue;
     }
 
-    R_len_t index_size = vec_size(VECTOR_ELT(indices, i));
+    R_len_t index_size = Rf_length(VECTOR_ELT(indices, i));
     out_size += index_size;
 
-    if (has_names) {
-      p_sizes[i] = index_size;
-    }
-
-    // Each element of `x` is recycled to its corresponding index's size
-    elt = vec_recycle(elt, index_size, args_empty);
-    SET_VECTOR_ELT(x, i, elt);
+    // Each element of `xs` is recycled to its corresponding index's size
+    x = vec_recycle(x, index_size, args_empty);
+    SET_VECTOR_ELT(xs, i, x);
   }
 
-  indices = PROTECT(vec_as_indices(indices, out_size, R_NilValue));
+  SEXP locs = PROTECT(vec_as_indices(indices, out_size, R_NilValue));
```

Comment:
I have a tough time naming this. We often use `locs` as an integer vector of locations, but this is a list-of-locs and we don't really have a good name for that. That was why I used `indices`

Reply:
hmm I think we use `loc` for an integer vector of location. In your vec-prefixes proposals, `locs` refers to a list of loc vectors.